label_studio/fsm/state_manager.py

```diff
@@ -0,0 +1,422 @@
+"""
+Core state management functionality for Label Studio.
+
+Provides high-performance state management with caching and batch operations
+that can be extended by Label Studio Enterprise with additional features.
+"""
+
+import logging
+from datetime import datetime
+from typing import TYPE_CHECKING, Any, Dict, List, Optional, Type
+
+from django.conf import settings
+from django.core.cache import cache
+from django.db import transaction
+from django.db.models import Model
+from fsm.models import BaseState
+from fsm.registry import get_state_model_for_entity
+
+# Avoid circular import
+if TYPE_CHECKING:
+    from fsm.transitions import BaseTransition
+
+logger = logging.getLogger(__name__)
+
+
+class StateManagerError(Exception):
+    """Base exception for StateManager operations"""
+
+    pass
+
+
+class InvalidTransitionError(StateManagerError):
+    """Raised when an invalid state transition is attempted"""
+
+    pass
+
+
+class StateManager:
+    """
+    Core state management system for Label Studio.
+
+    Provides the foundation for state management that can be extended
+    by Label Studio Enterprise with additional features like:
+    - Advanced caching strategies
+    - Bulk operations optimization
+    - Complex transition validation
+    - Enterprise-specific state models
+
+    Features:
+    - INSERT-only architecture with UUID7 for maximum performance
+    - Basic caching for current state lookups
+    - Simple state transitions with audit trails
+    - Extensible design for enterprise features
+    """
+
+    CACHE_TTL = getattr(settings, 'FSM_CACHE_TTL', 300)  # 5 minutes default
+    CACHE_PREFIX = 'fsm:current'
+
+    @classmethod
+    def get_cache_key(cls, entity: Model) -> str:
+        """Generate cache key for entity's current state"""
+        return f'{cls.CACHE_PREFIX}:{entity._meta.label_lower}:{entity.pk}'
+
+    @classmethod
+    def get_current_state(cls, entity: Model) -> Optional[str]:
+        """
+        Get current state with basic caching.
+
+        Args:
+            entity: The entity to get current state for
+
+        Returns:
+            Current state string or None if no states exist
+
+        Example:
+            task = Task.objects.get(id=123)
+            current_state = StateManager.get_current_state(task)
+            if current_state == 'COMPLETED':
+                # Task is finished
+                pass
+        """
+        cache_key = cls.get_cache_key(entity)
+
+        # Try cache first
+        cached_state = cache.get(cache_key)
+        if cached_state is not None:
+            logger.debug(f'Cache hit for {entity._meta.label_lower} {entity.pk}: {cached_state}')
+            return cached_state
+
+        # Query database using state model registry
+        state_model = get_state_model_for_entity(entity)
+        if not state_model:
+            logger.warning(f'No state model found for {entity._meta.model_name}')
```

Comment:
Maybe add something like "when attempting to get current state". Otherwise unclear where this is coming from